domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt

```diff
@@ -116,11 +133,51 @@ class AppStartupStateController @Inject constructor(
   }
 
   private fun computeAppStartupMode(onboardingState: OnboardingState): StartupMode {
-    return when {
-      hasAppExpired() -> StartupMode.APP_IS_DEPRECATED
-      onboardingState.alreadyOnboardedApp -> StartupMode.USER_IS_ONBOARDED
-      else -> StartupMode.USER_NOT_YET_ONBOARDED
+    // Return old logic if app and os feature flag is not enabled
+    if (!enableAppAndOsDeprecation.value) {
+      return when {
+        hasAppExpired() -> StartupMode.APP_IS_DEPRECATED
+        onboardingState.alreadyOnboardedApp -> StartupMode.USER_IS_ONBOARDED
+        else -> StartupMode.USER_NOT_YET_ONBOARDED
+      }
     }
+
+    val deprecationDataProvider = deprecationController.getDeprecationDatabase()
+
+    var deprecationDatabase = DeprecationResponseDatabase.newBuilder().build()
+
+    runBlocking {
+      deprecationDataProvider.retrieveData().transform {
+        deprecationDatabase = it
+      }
+    }
+
+    val appVersionCode = Build.VERSION.SDK_INT
+
+    val osIsDeprecated = lowestSupportedApiLevel.value > appVersionCode &&
+      deprecationDatabase.osDeprecationResponse.deprecatedVersion != appVersionCode
+    val appUpdateIsAvailable = optionalAppUpdateVersionCode.value > appVersionCode ||
+      forcedAppUpdateVersionCode.value > appVersionCode
+
+    if (onboardingState.alreadyOnboardedApp) {
+      if (osIsDeprecated) {
+        return StartupMode.OS_IS_DEPRECATED
+      }
+
+      if (appUpdateIsAvailable) {
+        if (forcedAppUpdateVersionCode.value > appVersionCode) {
+          return StartupMode.APP_IS_DEPRECATED
+        }
+
+        if (deprecationDatabase.appDeprecationResponse.deprecatedVersion !=
+          optionalAppUpdateVersionCode.value
+        ) {
+          return StartupMode.OPTIONAL_UPDATE_AVAILABLE
+        }
+      }
+
```

Comment:
This can be extracted to a public api in the `DeprecationController`

Reply:
I have migrated this logic to the specified class.